### ISSUE TITLE
Caching question. metadata, response and etc.

### ISSUE THREAD

**Issue by sergiyshutov** [kaltura/nginx-vod-module#267]
Dear erankor. Once again question :)

In my server i have 32gb ram, and always free.
What about this variables vod module, can you please answer me?

vod_metadata_cache metadata_cache 4000m;
vod_response_cache response_cache 512m;

vod_metadata_cache seems i need bigest value? Like 30g? :D :D Or i wrong?

----------------------------------------
And one more

`location ~ ^/hls/(?<secure>[\w\-=]+,\d+)(?<file>/.*\.mp4)(.*)$ {
            secure_link $secure;
            secure_link_md5 $secure_link_expires.$file.SECURETEXT;

            if ($secure_link = "")  { return 403; }
            if ($secure_link = "0") { return 403; }

            alias /home/www/vod/video/$file;

            vod hls;
            vod_segment_duration 10000;
            vod_bootstrap_segment_durations 2000;
            vod_bootstrap_segment_durations 2000;
            vod_bootstrap_segment_durations 2000;
            vod_bootstrap_segment_durations 2000;
            vod_bootstrap_segment_durations 2000;
            vod_bootstrap_segment_durations 2000;
            vod_bootstrap_segment_durations 2000;
            vod_bootstrap_segment_durations 2000;
            vod_bootstrap_segment_durations 2000;
            vod_bootstrap_segment_durations 2000;

            add_header Access-Control-Allow-Headers '*';
            add_header Access-Control-Expose-Headers 'Server,range,Content-Length,Content-Range';
            add_header Access-Control-Allow-Methods 'GET, HEAD, OPTIONS';
            add_header Access-Control-Allow-Origin '*';

            expires max;
        }`

So, url video file is:
/hls/v12WgpYUCHcUccyQZVZ5OA,1459676381/2016/4/7wzgx0wYj5QHdShmqa9g.mp4

But this url always changing on new request. Like, i refresh page and url is v12WgpYUCHcUccyQZVZ5OA,1459676381 i refresh again and url is v242vsjadkaslsh3423424.... something like that. And i need download .ts files again. Maybe you can tell me what i can to do? 
And how this affect cache variables like vod_metadata_cache and etc.?

Big thanks.

**Comment by erankor** [comment]
1. Regarding cache sizes - 512m for the response cache is probably way too much, this cache only holds manifest responses, the module does not cache video segments, these are assumed to be cache in some proxy servers located closer to the end users. Btw, the reason I added this cache is for cases in which there is tokenization, so the final manifest with the tokens can't be cached for long, but the original manifest that doesn't have tokens can be.
2. Regarding the metadata cache - this cache should indeed get the largest portion, I can't really tell you how much is a good value for you. I suggest you will look at the performance counters that nginx-vod-module maintains (configure `vod_performance_counters` and add some location with `vod_status`), you will see there how much space is allocated by each cache, and also the hit/miss values. If you see that the metadata cache is fully used, and the hit/miss ratio is not great (let's say, lower than 80%) you should probably increase the cache size.
3. Regarding the URLs - are you saying that in your setup the URLs for the videos are always changing ? if it's only the query string that changes then it's fine, but if the path of the URL changes then it's really a problem in terms of caching, since the cache key is calculated according to the domain + url path. If the path is random, you will get a very poor cache hit ratio.

**Comment by sergiyshutov** [comment]
Dear erankor. Thank you very much. I understand now :)

About your 3 answer. Can you please explain how i can use your nginx-secure-token-module? I really don't understand how to use this. Because doc explain is very small. How to use without any akamai and etc. I dont use CDN. And how you generate key? Yes, you have example encryptUrl.php how to generate link key. But again, what is **encrypted part** what is **encryption iv** and etc.

**Comment by erankor** [comment]
nginx-secure-token-module embeds tokens in manifest responses, the idea is that for HLS for example, you create a token for index.m3u8 in your application, but then the token has to propagate to the segment requests (seg-1.ts, etc.), that is what this module does.
Other than that, it has a separate feature of encrypting parts of the URLs, this is something we needed for some very specific use case. The encryptUrl script you referred to is related to this functionality (not related to tokens).
If you are not working with a CDN, then other than generating the tokens in your app, and making them propagate to the segments (using nginx-secure-token-module) you'd also need something to validate the tokens, and fail the request if they are expired / badly signed etc. There's another module we maintain https://github.com/kaltura/nginx-akamai-token-validate-module that does that part. You can think of it as a module similar to the nginx builtin secure link module, only that it uses Akamai token format.
So, to summarize, even though you don't use Akamai, you can decide to use Akamai token format, and have:
1. nginx-akamai-token-validate-module - validate the tokens
2. nginx-secure-token-module - make the token propagate from the manifest request to the segment requests
3. generate the Akamai tokens in your application, we use this PHP code for that: https://github.com/kaltura/server/blob/Kajam-11.12.0/alpha/apps/kaltura/lib/storage/urlTokenizers/kAkamaiSecureHDUrlTokenizer.php
